vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,785 @@
+#include "conditional_contextual_bandit.h"
+#include "reductions.h"
+#include "example.h"
+#include "global_data.h"
+#include "cache.h"
+#include "vw.h"
+#include "interactions.h"
+#include "label_dictionary.h"
+#include "cb_adf.h"
+#include "cb_algs.h"
+
+#include <numeric>
+#include <algorithm>
+#include <unordered_set>
+#include <unordered_map>
+
+using namespace LEARNER;
+using namespace VW;
+using namespace VW::config;
+
+
+struct ccb
+{
+  example* shared;
+  std::vector<example*> actions, decisions;
+  uint32_t chosen_action_index;
+  std::unordered_map<uint32_t, uint32_t> origin_index;
+  CB::cb_class cb_label, default_cb_label;
+  std::unordered_set<uint32_t> exclude_list, include_list;
+  CCB::decision_scores_t decision_scores;
+  std::vector<std::string>* original_interactions;
+};
+
+namespace CCB
+{
+static constexpr uint32_t SHARED_EX_INDEX = 0;
+static constexpr uint32_t TOP_ACTION_INDEX = 0;
+
+void clear_all(ccb& data)
+{
+  data.shared = nullptr;
+  data.actions.clear();
```

Comment:
This could be a problem if we're dealing with a large enough set of actions/decisions that calling `clear` actually relocates the underlying array.

Reply:
The size of the vector should be sufficiently small for it not to matter here, but I agree this is not great. Would be good if there was a way to clear it and guarantee the memory is not deallocated

Reply:
v_array?

Reply:
I don't think using v_array is necessary here. v_array still reallocates as well after a number of clears